README.md
yenc-freddie 0.3
================

Disclaimer: 
-----------
Not my software, just figured I would put it on github, since
it would appear that this is somewhat hard to find.

Use: 
----
This a fairly simple module, it provide only raw yEnc encoding/decoding with
builitin crc32 calculation.
Header parsing, checkings and yenc formatting are left to you (see examples 
directory for possible implementations). The interface was originally intended 
to be similar to the uu module from Python standard library.
Version 0.2 changed a bit the previous (0.1)  behaviour, but it should be
more usable and flexible (now you can encode/decode from and to standard
input and output and use filenames instead of file objects as arguments
for encode() and decode() ). See CHANGES file for details.
Version 0.3 doesn't introduce anything new, just a bugfix a some internals
changes

good yenc module for use with newsmangler, the yenc-vanilla one included with newsmangler is no good.

```bash
How to install:

1.  python setup.py build
2.  sudo python setup.py install
```
```bash
How to use with newsmangler:
1.  edit newsmangler/newsmangler/yenc.py
2.  change HAVE_YENC_FRED = True
3.  Go to bottom of yenc.py
4.  Change HAVE_YENC_FRED = ('Freddie mod' in _yenc.__doc__)
		to
	HAVE_YENC_FRED = ('Freddie' in _yenc.__doc__)
```






For more information check out README-ORIGINAL
----------------------------------------------



CHANGES

$Id: CHANGES,v 1.2 2004/02/18 23:05:05 cvs Exp $

version: 0.1
- Initial release.

version: 0.2
- Included patch from Michael Muller <mmuller@enduden.com> that fixes some 
possible buffer overflows, and some yEnc incompatibilities.
- A few enanchments in C code, mainly cosmetic (it should be a little faster
in encoding).
- Files opened in O_RDWR mode are now handled properly.
- yenc.i interface adapted to latest SWIG releases.
- Passing a file opened in a wrong mode now raises a ValueError (rather than
an AssertionError).
- encode()/decode() now accepts filenames and stdin/stdout as arguments.
- Reaching EOF while encoding/decoding no more raises an exception (it was
redundant since the number of encoded/decoded bytes were already returned),
this could break some implementations.
- Passing 0 to the encode o decode functions causes file to be encoded/decoded
up to EOF (this is made possible by the previous change).
 
version: 0.3
- Encoded lines now are correctly terminated with CRLF.
- Encoded lines size now is 128 chars.
- using fread() instead of fgets in decode() function (much faster).
- SWIG isn't used anymore for wrapper generation.
- freddie: added escapedots parameter to encode_string, adds the required
  extra period at the start of line for usenet posts.


test/test.py
#!/usr/bin/env python
##=============================================================================
 #
 # Copyright (C) 2003, 2004 Alessandro Duca
 #
 # This program is free software; you can redistribute it and/or
 # modify it under the terms of the GNU General Public License
 # as published by the Free Software Foundation; either version 2
 # of the License, or (at your option) any later version.
 # 
 # This program is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
 # 
 # See the GNU General Public License for more details.
 # 
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 #
 #=============================================================================
 #
 # $Id: test.py,v 1.2 2004/02/18 23:11:46 cvs Exp $
 # 
##=============================================================================

import yenc
import os
import sys
import time
from binascii import crc32
from stat import *

def main():
	os.system("dd if=/dev/urandom of=sampledata bs=1M count=1")
	file_in = open("sampledata","r")
	file_out = open("sampledata.out","w")
	data = file_in.read()
	file_in.seek(0,0)
	size = len(data)
	crc = hex(crc32(data))[2:]
	print "initial file crc: %s dec: %d" % (crc, crc32(data))
	del(data)
	print "Starting encoding test:"
	startime = time.time()
	bytes_out, crc_out = yenc.encode(file_in, file_out, size)
	secs = time.time() - startime
	print "Bytes out:", bytes_out, "output crc:", crc_out
	if crc != crc_out:
		print "crc error, 1Mb block encoding"
		sys.exit(1)
	elif bytes_out != size:
		print "size error, 1Mb block encoding"
		sys.exit(1)
	else:
		print "crc and size, 1Mb block encoding ok. %f secs." % secs

	file_in.close()
	file_out.close()
	os.unlink("sampledata")
	file_in = open("sampledata.out","r")
	file_out = open("sampledata","w")
	print "Starting decoding test:"
	startime = time.time()
	bytes_out, crc_out = yenc.decode(file_in, file_out, size)
	secs = time.time() - startime
	print "Bytes out:", bytes_out, "output crc:", crc_out
	if crc != crc_out:
		print "crc error, 1Mb block decoding"
		sys.exit(1)
	elif bytes_out != size:
		print "size error, 1Mb block decoding"
		sys.exit(1)
	else:
		print "crc and size, 1Mb block decoding ok. %f secs." % secs
	os.unlink("sampledata")
	os.unlink("sampledata.out")

if __name__ == "__main__":
	main()


lib/yenc.py
##=============================================================================
 #
 # Copyright (C) 2003, 2004 Alessandro Duca
 #
 # This program is free software; you can redistribute it and/or
 # modify it under the terms of the GNU General Public License
 # as published by the Free Software Foundation; either version 2
 # of the License, or (at your option) any later version.
 # 
 # This program is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
 # 
 # See the GNU General Public License for more details.
 # 
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 #
 #=============================================================================
 #
 # $Id: yenc.py,v 1.2 2004/02/18 23:08:15 cvs Exp $
 # 
##=============================================================================


import sys
from cStringIO import StringIO
import _yenc

E_ERROR		= 64
E_CRC32		= 65
E_PARMS		= 66

class Error(Exception):
	""" 	Class for specific yenc errors
	"""

	def __init__(self, value="", code=E_ERROR):
		self.value = value
		
	def __str__(self):
		return "yenc.Error: %s\n" % self.value, self.value


def _checkArgsType(file_in, file_out, bytes):
	""" 	Internal checkings, not to be used from outside this module
	"""
	
	if bytes < 0: raise Error("No. of bytes can't be negative", E_PARMS)
	if type(file_in) == str:
		if file_in == "-":
			if bytes == 0: raise Error("No. of bytes is 0 or not \
				specified while reading from stdin", E_PARMS)
			file_in = sys.stdin
		else: file_in = open(file_in,"rb")
	if type(file_out) == str:
		if file_out == "-": file_out = sys.stdout
		else: file_out = open(file_out,"wb")
	return file_in, file_out, bytes


def encode(file_in, file_out, bytes=0):
	"""	encode(file_in, file_out, bytes=0): write "bytes" encoded bytes from
		file_in to file_out, if "bytes" is 0 encodes bytes until EOF
	"""
	
	file_in, file_out, bytes = _checkArgsType(file_in, file_out, bytes)
	encoded, crc32 = _yenc.encode(file_in, file_out, bytes)
	crc_hex = hex(crc32)[2:]
	return encoded, crc_hex


def decode(file_in, file_out, bytes=0, crc_in=""):
	""" 	decode(file_in, file_out, bytes=0): write "bytes" decoded bytes from
		file_in to file_out, if "bytes" is 0 decodes bytes until EOF
	"""
	
	file_in, file_out, bytes = _checkArgsType(file_in, file_out, bytes)
	decoded, crc32 = _yenc.decode(file_in, file_out, bytes)
	crc_hex = hex(crc32)[2:]
	if crc_in and not cmp(crc_hex,crc_in.lower()):
		raise Error("crc32 error", E_CRC32)
	else:
		return decoded, crc_hex


class Encoder:
	""" 	class Encoder: facility class for encoding one string at time
	"""
	
	def __init__(self, output_file = None):
		self._buffer = StringIO()
		self._column = 0
		self._output_file = output_file
		self._crc = -1
		self._encoded = 0
		self._terminated = 0

	def __del__(self):
		if(self._output_file):
			self._output_file.flush()
			self._output_file.close()
	
	def feed(self, data):
		"""	Encode some data and write the encoded data 
			into the internal buffer
		"""
		if self._terminated:
			raise IOError("Encoding already terminated")
		encoded, self._crc, self._column = _yenc.encode_string(data, 
							self._crc, self._column)
		self._encoded = self._encoded + len(encoded)
		self._buffer.write(encoded)
		return len(encoded)
	
	def terminate(self):
		"""	Appends the terminal CRLF sequence to the encoded data
		"""
		self._terminated = 1
		self._buffer.write("\r\n")
	
	def flush(self):
		"""	Writes the content of the internal buffer on the file
			passed as argument to the constructor
		"""
		if self._output_file:
			self._output_file.write(self._buffer.getvalue())
			self._buffer = StringIO()
		else:
			raise ValueError("Output file is 'None'")
	
	def getEncoded(self):
		"""	Returns the data in the internal buffer
		"""
		if not self._output_file:
			return self._buffer.getvalue()
		else:
			raise ValueError("Output file is not 'None'")
	
	def getSize(self):
		"""	Returns the total number of encoded bytes (not the size of
			the buffer)
		"""
		return self._encoded

	def getCrc32(self):
		"""	Returns the calculated crc32 string for the clear
			encoded data
		"""
		return "%08x" % (self._crc ^ -1)


class Decoder:
	""" class Decoder: facility class for decoding one string at time
	"""
	
	def __init__(self, output_file = None):
		self._buffer = StringIO()
		self._escape = 0
		self._output_file = output_file
		self._crc = -1
		self._decoded = 0
	
	def __del__(self):
		if(self._output_file):
			self._output_file.flush()
			self._output_file.close()
	
	def feed(self, data):
		"""	Encode some data and write the encoded data 
			into the internal buffer
		"""
		decoded, self._crc, self._escape = _yenc.decode_string(data, self._crc, self._escape)
		self._decoded = self._decoded + len(decoded)
		self._buffer.write(decoded)
		return len(decoded)
	
	def flush(self):
		"""	Writes the content of the internal buffer on the file
			passed as argument to the constructor
		"""
		if self._output_file:
			self._output_file.write(self._buffer.getvalue())
			self._buffer = StringIO()
		else:
			raise ValueError("Output file is 'None'")
	
	def getDecoded(self):
		"""	Returns the data in the internal buffer
		"""
		if not self._output_file:
			return self._buffer.getvalue()
		else:
			raise ValueError("Output file is not 'None'")

	def getSize(self):
		"""	Returns the total number of decoded bytes (not the size of
			the buffer)
		"""
		return self._decoded
	
	def getCrc32(self):
		"""	Returns the calculated crc32 string for the decoded data
		"""
		return "%08x" % (self._crc ^ -1)



src/_yenc.h
/*=============================================================================
 *
 * Copyright (C) 2003, 2004 Alessandro Duca
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
 * 
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 *=============================================================================
 *
 * $Id: _yenc.h,v 1.2 2004/02/18 23:06:14 cvs Exp $
 * 
 */



#include <Python.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>

/* Constants			*/
#define	LINESIZE	128

/* BLOCK defines the size of the input buffer used while encoding data.
 */
#define BLOCK		65536

/* LONGBUFF: ((2 * BLOCK) / 128 + 2 ) * 128
 * In the worst case we will escape every byte thus doubling output's size,
 * since we write out lines of LINESIZE characters, we must add 2 more bytes for
 * CRLF sequence at the end of each line.
 * This is the maximum encoded size of BLOCK bytes.
 */
#define LONGBUFF	( 2 * BLOCK / LINESIZE + 1) * ( LINESIZE + 2 ) 

#define SMALLBUFF 	512

#define ZERO		0x00
#define CR		0x0d
#define	LF		0x0a
#define	ESC		0x3d
#define TAB		0x09
#define SPACE	0x20
#define DOT		0x2E

#define E_MODE		1
#define E_EOF 		2
#define E_IO		3

#define E_MODE_MSG	"Invalide mode for '*file' arguments"
#define E_IO_MSG	"I/O Error"


/* Customized types		*/
typedef unsigned long uLong;
typedef unsigned char Byte;
typedef int Bool;

/* Functions */
PyObject* encode_file(PyObject*, PyObject*, PyObject*);
PyObject* decode_file(PyObject*, PyObject*, PyObject*);
PyObject* encode_string(PyObject* ,PyObject* ,PyObject*);
PyObject* decode_string(PyObject* ,PyObject* , PyObject*);
void init_yenc(void);





examples/ydecode.py
#!/usr/bin/env python
##=============================================================================
 #
 # Copyright (C) 2003, 2004 Alessandro Duca
 #
 # This program is free software; you can redistribute it and/or
 # modify it under the terms of the GNU General Public License
 # as published by the Free Software Foundation; either version 2
 # of the License, or (at your option) any later version.
 # 
 # This program is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
 # 
 # See the GNU General Public License for more details.
 # 
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 #
 #=============================================================================
 #
 # $Id: ydecode.py,v 1.2 2004/02/18 23:10:11 cvs Exp $
 # 
##=============================================================================

import yenc
import os
import sys
import re

NAME_RE 	= re.compile(r"^.*? name=(.+?)\r\n$")
LINE_RE 	= re.compile(r"^.*? line=(\d{3}) .*$")
SIZE_RE 	= re.compile(r"^.*? size=(\d+) .*$")
CRC32_RE	= re.compile(r"^.*? crc32=(\w+)")

def main():
	if len(sys.argv) > 1:
		file_in = open(sys.argv[1],"r")
	else:
		file_in	= sys.stdin
	while 1:
		line = file_in.readline()
		if line.startswith("=ybegin "):
			try:
				name, size = NAME_RE.match(line).group(1), int(SIZE_RE.match(line).group(1))
				m_obj = CRC32_RE.match(line)
				if m_obj:
					head_crc = m_obj.group(1)
			except re.error, e:
				sys.stderr.write("err-critical: malformed =ybegin header\n")
				sys.exit(1)
			break
		elif not line:
			sys.stderr.write("err-critical: no valid =ybegin header found\n")
			sys.exit(1)
	file_out = open(name,"w")
	try:
		dec, dec_crc = yenc.decode(file_in, file_out, size)
	except yenc.Error, e:
		sys.stderr.write(str(e) + '\n')
		sys.exit(1)
	head_crc = trail_crc = tmp_crc = ""
	garbage	= 0
	for line in file_in.read().split("\r\n"):
		if line.startswith("=yend "):
			try:	
				size = int( SIZE_RE.match(line).group(1) )
				m_obj = CRC32_RE.match(line)
				if m_obj:
					trail_crc = m_obj.group(1)
			except re.error, e:
				sys.stderr.write("err: malformed =yend trailer\n")
			break
		elif not line:
			continue
		else:
			garbage = 1
	else:
		sys.stderr.write("warning: couldn't find =yend trailer\n")
	if garbage:
		sys.stderr.write("warning: garbage before =yend trailer\n")
	if head_crc:
		tmp_crc = head_crc.lower()
	elif trail_crc:
		tmp_crc = trail_crc.lower()
	else:
		sys.exit(0)
	if cmp(tmp_crc,dec_crc):
		sys.stderr.write("err: header: %s dec: %s CRC32 mismatch\n" % (tmp_crc,dec_crc) )
		sys.exit(1)
	else:
		sys.exit(0)

if __name__ == "__main__":
	main()


src/_yenc.c
/*=============================================================================
 *
 * Copyright (C) 2003, 2004 Alessandro Duca
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
 * 
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 *=============================================================================
 *
 * $Id: _yenc.c,v 1.2 2004/02/18 23:06:14 cvs Exp $
 * 
 */

#include "_yenc.h"

/* Typedefs */
typedef struct {
	int crc;
	uLong bytes;
} Crc32;

/* Declarations */
static int crc_tab[256] = {
0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f, 
0xe963a535, 0x9e6495a3, 0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988, 
0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91, 0x1db71064, 0x6ab020f2, 
0xf3b97148, 0x84be41de, 0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7,
0x136c9856, 0x646ba8c0, 0xfd62f97a, 0x8a65c9ec, 0x14015c4f, 0x63066cd9, 
0xfa0f3d63, 0x8d080df5, 0x3b6e20c8, 0x4c69105e, 0xd56041e4, 0xa2677172, 
0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b, 0x35b5a8fa, 0x42b2986c, 
0xdbbbc9d6, 0xacbcf940, 0x32d86ce3, 0x45df5c75, 0xdcd60dcf, 0xabd13d59,
0x26d930ac, 0x51de003a, 0xc8d75180, 0xbfd06116, 0x21b4f4b5, 0x56b3c423, 
0xcfba9599, 0xb8bda50f, 0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924, 
0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d, 0x76dc4190, 0x01db7106, 
0x98d220bc, 0xefd5102a, 0x71b18589, 0x06b6b51f, 0x9fbfe4a5, 0xe8b8d433,
0x7807c9a2, 0x0f00f934, 0x9609a88e, 0xe10e9818, 0x7f6a0dbb, 0x086d3d2d, 
0x91646c97, 0xe6635c01, 0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e, 
0x6c0695ed, 0x1b01a57b, 0x8208f4c1, 0xf50fc457, 0x65b0d9c6, 0x12b7e950, 
0x8bbeb8ea, 0xfcb9887c, 0x62dd1ddf, 0x15da2d49, 0x8cd37cf3, 0xfbd44c65,
0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2, 0x4adfa541, 0x3dd895d7, 
0xa4d1c46d, 0xd3d6f4fb, 0x4369e96a, 0x346ed9fc, 0xad678846, 0xda60b8d0, 
0x44042d73, 0x33031de5, 0xaa0a4c5f, 0xdd0d7cc9, 0x5005713c, 0x270241aa, 
0xbe0b1010, 0xc90c2086, 0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4, 0x59b33d17, 0x2eb40d81, 
0xb7bd5c3b, 0xc0ba6cad, 0xedb88320, 0x9abfb3b6, 0x03b6e20c, 0x74b1d29a, 
0xead54739, 0x9dd277af, 0x04db2615, 0x73dc1683, 0xe3630b12, 0x94643b84, 
0x0d6d6a3e, 0x7a6a5aa8, 0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1,
0xf00f9344, 0x8708a3d2, 0x1e01f268, 0x6906c2fe, 0xf762575d, 0x806567cb, 
0x196c3671, 0x6e6b06e7, 0xfed41b76, 0x89d32be0, 0x10da7a5a, 0x67dd4acc, 
0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5, 0xd6d6a3e8, 0xa1d1937e, 
0x38d8c2c4, 0x4fdff252, 0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55, 
0x316e8eef, 0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236, 
0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f, 0xc5ba3bbe, 0xb2bd0b28, 
0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7, 0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d,
0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a, 0x9c0906a9, 0xeb0e363f, 
0x72076785, 0x05005713, 0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38, 
0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242, 
0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69, 
0x616bffd3, 0x166ccf45, 0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2, 
0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc, 
0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693, 
0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94, 
0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
};
static char* argnames[] = {"infile", "outfile", "bytes", NULL};

/* Function declarations */
static void crc_init(Crc32 *, int);
static void crc_update(Crc32 *, int);
static Bool readable(FILE *);
static Bool writable(FILE *);
void init_yenc(void);
static int encode_buffer(char *, char *, int, Crc32*, int *, int);
static int decode_buffer(char *, char *, int, Crc32*, Bool *);
PyObject* decode_string(PyObject* ,PyObject* ,PyObject* );

/* Python API requirements */
static char encode_doc[] = "encode(input_file, output_file, <size>)";
static char decode_doc[] = "decode(input_file, output_file, <size>)";
static char encode_string_doc[] = "encode_string(input_string, output_string, startcolumn, escapedots)";
static char decode_string_doc[] = "decode_string(input_string, output_string, escape)";
static PyMethodDef funcs[] = {
		{"encode", (PyCFunction) encode_file, METH_KEYWORDS | METH_VARARGS, encode_doc},
		{"decode", (PyCFunction) decode_file, METH_KEYWORDS | METH_VARARGS, decode_doc},
		{"encode_string", (PyCFunction) encode_string, METH_KEYWORDS | METH_VARARGS, encode_string_doc},
		{"decode_string", (PyCFunction) decode_string, METH_KEYWORDS | METH_VARARGS, decode_string_doc},
		{NULL, NULL, 0, NULL}
};
/* Function definitions */
static void crc_init(Crc32 *crc, int value)
{
	crc->crc = value;
	crc->bytes = 0UL;
}

static void crc_update(Crc32 *crc ,int c)
{
	crc->crc=crc_tab[(crc->crc^(c&0x000000ffL))&0xffL]^((crc->crc>>8L)&0xffffffL);
	crc->bytes++;
}

/*
 * Todo: provide alternatives for this to work on win32
 */
static Bool writable(FILE *file)
{
	int mode = fcntl(fileno(file),F_GETFL) & O_ACCMODE;
	return (mode == O_WRONLY) || (mode == O_RDWR);
}

static Bool readable(FILE *file)
{
	int mode = fcntl(fileno(file),F_GETFL) & O_ACCMODE;
	return (mode == O_RDONLY) || (mode == O_RDWR);
}
/*
 * 
 */

static int encode_buffer(
		char *input_buffer, 
		char *output_buffer, 
		int bytes, 
		Crc32 *crc, 
		int *col,
		int escapedots
		)
{
	int encoded;
	int in_ind;
	int out_ind;
	Byte byte;
	
	out_ind = 0;
	for(in_ind=0; in_ind < bytes; in_ind++) {
		byte = (Byte)( input_buffer[in_ind] + 42 );
		crc_update(crc, input_buffer[in_ind]);
		switch(byte){
			case ZERO:
			case LF:
			case CR:
			case ESC:
				goto escape_string;
			case TAB:
			case SPACE:
				if(*col == 0 || *col == LINESIZE-1) {
					goto escape_string;
				}
			case DOT:
				if(escapedots && *col == 0) {
					output_buffer[out_ind++] = DOT;
					(*col)++;
				}
			default:
				goto plain_string;
		}
		escape_string:
		byte = (Byte)(byte + 64);
		output_buffer[out_ind++] = ESC;
		(*col)++;
		plain_string:
		output_buffer[out_ind++] = byte;
		(*col)++;
		encoded++;
		if(*col >= LINESIZE) {
			output_buffer[out_ind++] = CR;
			output_buffer[out_ind++] = LF;
			*col = 0;
		}
	}
	return out_ind;
}

PyObject* encode_file(
		PyObject* self,
		PyObject* args,
		PyObject* kwds
		)
{
	Byte read_buffer[BLOCK];
	Byte write_buffer[LONGBUFF];
	uLong encoded	= 0;
	int col		= 0;
	int read_bytes;
	int in_ind;
	int encoded_bytes;
	uLong bytes = 0;
	Crc32 crc;

	FILE *infile = NULL, *outfile = NULL;
	PyObject *Py_infile = NULL, *Py_outfile = NULL;
	
	if(!PyArg_ParseTupleAndKeywords(args, kwds, "O!O!|l", argnames, \
				&PyFile_Type, &Py_infile, \
				&PyFile_Type, &Py_outfile, \
				&bytes)) return NULL;

	infile = PyFile_AsFile(Py_infile);
	outfile = PyFile_AsFile(Py_outfile);
	
	if(!readable(infile) || !writable(outfile) ) {
		return PyErr_Format(PyExc_ValueError, "file objects not writeable/readable");
	} 
	
	crc_init(&crc, -1);

	while(encoded < bytes || bytes == 0){
		if( bytes && (bytes - encoded) < BLOCK) {
			in_ind = bytes - encoded;
		} else {
			in_ind = BLOCK;
		}
		read_bytes = fread(&read_buffer, 1, in_ind, infile);
		if(read_bytes < 1) break;
		encoded_bytes = encode_buffer(&read_buffer[0], &write_buffer[0], read_bytes, &crc, &col, 0);
		if(fwrite(&write_buffer, 1, encoded_bytes, outfile) != encoded_bytes) {
			break;
		}
		encoded += read_bytes;
	}
	if(ferror(infile) || ferror(outfile)) {
		return PyErr_Format(PyExc_IOError, "I/O Error while encoding");
	}
	if(col > 0) {
		fputc( CR, outfile );
		fputc( LF, outfile );
	}
	fflush(outfile);
	return Py_BuildValue("(l,i)",encoded,(unsigned int) (crc.crc ^ 0xFFFFFFFF));
}

static int decode_buffer(
		char *input_buffer, 
		char *output_buffer, 
		int bytes, 
		Crc32 *crc, 
		Bool *escape
		)
{
	int read_ind;
	int decoded_bytes;
	Byte byte;
	
	decoded_bytes = 0;
	for(read_ind = 0; read_ind < bytes; read_ind++) {
		byte = input_buffer[read_ind];
		if(*escape) {
			byte = (Byte)( byte - 106 );
			*escape = 0;
		} else if(byte == ESC) {
			*escape = 1;
			continue;
		} else if(byte == LF || byte == CR) {
			continue;
		} else {
			byte = (Byte)( byte - 42 );
		}
		output_buffer[decoded_bytes] = byte;
		decoded_bytes++;
		crc_update(crc, byte);
	}
	return decoded_bytes;
}

PyObject* encode_string(
		PyObject* self, 
		PyObject* args, 
		PyObject* kwds
		)
{
	PyObject *Py_input_string;
	PyObject *Py_output_string;
	PyObject *retval;
	
	Byte *input_buffer = NULL;
	Byte *output_buffer = NULL;
	int crc_value = -1;
	int input_len = 0;
	int output_len = 0;
	int col = 0;
	int escapedots = 0;
	Crc32 crc;
	
	static char *kwlist[] = { "string", "crc32", "column", "escapedots", NULL };
	if(!PyArg_ParseTupleAndKeywords(args, 
				kwds,
				"O!|iii", 
				kwlist,
				&PyString_Type,
				&Py_input_string, 
				&crc_value,
				&col,
				&escapedots
				)) 
		return NULL;
	crc_init(&crc, crc_value);
	input_len = PyString_Size(Py_input_string);
	input_buffer = PyString_AsString(Py_input_string);
	output_buffer = (char *) malloc(( 2 * input_len / LINESIZE + 1 ) * ( LINESIZE + 2 ));
	output_len = encode_buffer(input_buffer, output_buffer, input_len, &crc, &col, escapedots);
	Py_output_string = PyString_FromStringAndSize(output_buffer, output_len);
	retval = Py_BuildValue("(S,i,i)", Py_output_string, crc.crc, col);

	free(output_buffer);
	Py_DECREF(Py_output_string);
	
	return retval;
}

PyObject* decode_file(
		PyObject* self, 
		PyObject* args, 
		PyObject* kwds
		)
{
	Byte read_buffer[BLOCK];
	Byte write_buffer[LONGBUFF];
	uLong decoded = 0;
	int decoded_bytes;
	int read_bytes;
	uLong read_max;
	
	Bool escape = 0;
	uLong bytes = 0;
	Crc32 crc;

	FILE *infile = NULL, *outfile = NULL;
	PyObject *Py_infile = NULL, *Py_outfile = NULL;
	
	if(!PyArg_ParseTupleAndKeywords(args, kwds, "O!O!|l", argnames, \
				&PyFile_Type, &Py_infile, \
				&PyFile_Type, &Py_outfile, \
				&bytes)) return NULL;

	infile = PyFile_AsFile(Py_infile);
	outfile = PyFile_AsFile(Py_outfile);

	if(!readable(infile) || !writable(outfile)) {
		return PyErr_Format(PyExc_ValueError,
				"file objects not writeable/readable");
	} 

	crc_init(&crc, -1);
	
	while(decoded < bytes || bytes == 0){
		if(bytes){
			read_max=(bytes-decoded)<BLOCK?(bytes-decoded):BLOCK;
		} else {
			read_max=BLOCK;
		};
		read_bytes = fread((Byte *)&read_buffer, 1, read_max, infile);
		if(read_bytes == 0) break;
		decoded_bytes = decode_buffer(&read_buffer[0],
				&write_buffer[0],read_bytes, &crc, &escape);
		if(fwrite(&write_buffer[0],1,decoded_bytes,outfile)!=decoded_bytes){
			break;
		}
		decoded += decoded_bytes;
	}
	if(ferror(infile) || ferror(outfile)) {
		return PyErr_Format(PyExc_IOError, "I/O Error while decoding");
	}
	fflush(outfile);
	return Py_BuildValue("(l,i)", 
			decoded,(unsigned int) (crc.crc ^ 0xFFFFFFFF));
}

PyObject* decode_string(
		PyObject* self, 
		PyObject* args, 
		PyObject* kwds
		)
{
	PyObject *Py_input_string;
	PyObject *Py_output_string;
	PyObject *retval;
	
	char *input_buffer = NULL;
	char *output_buffer = NULL;
	long crc_value = -1L;
	int input_len = 0;
	int output_len = 0;
	int escape = 0;
	Crc32 crc;
	
	static char *kwlist[] = { "string", "crc32", "escape", NULL };
	if(!PyArg_ParseTupleAndKeywords(args, 
				kwds,
				"O!|ii", 
				kwlist,
				&PyString_Type,
				&Py_input_string, 
				&crc_value,
				&escape
				)) 
		return NULL;
	crc_init(&crc, crc_value);
	input_len = PyString_Size(Py_input_string);
	input_buffer = PyString_AsString(Py_input_string);
	output_buffer = (char *) malloc( input_len );
	output_len = decode_buffer(input_buffer, output_buffer, input_len, &crc, &escape);
	Py_output_string = PyString_FromStringAndSize(output_buffer, output_len);
	retval = Py_BuildValue("(S,i,i)", Py_output_string, crc.crc, escape);
	
	free(output_buffer);
	Py_DECREF(Py_output_string);

	return retval;
}


void init_yenc()
{
	Py_InitModule3("_yenc", funcs, "Raw yenc operations (mangled by Freddie)");
}

